<div align="center" style="color: #000099"> <h1><b>Climate Change Belief Analysis 2022</b></h1>
    <h4> By </h4>
    <h2>Datafluent Inc.</h2>
</div>

## About Datafluent Inc.
In this era of `Information Technology (IT)`, the variety and quantity of data generated on a daily basis is unprecedented. Data, which is sometimes referred to as the new oil, holds invaluable insights that can make the world a better place and businesses with growth mindsets have started exploiting this invaluable resource to keep thriving in the business world.

Behold **`Datafluent Inc.!`** We are an organisation that enjoys squeezing out every bit of actionable insights from a variety of raw facts and communicating such insights in digestible chunks to our stakeholders. We are a team of five seasoned professionals that excelled in various disciplines. 

## Meet the Team

<div align="center" style="width: 900px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="The team.jpg"
     alt="The team"
     style="float: center; padding-bottom=0.5em"
     width=900px/>
</div>

## Project Overview
Industrialization is the enabler of modern economic growth and development. However, this comes at the cost of emitting greenhouse gases that contribute, negatively, to climate change and ultimately global warming. Governments, eco-conscious organizations, and civil societies around the world are constantly exploring ways to reduce their carbon footprints. Despite many indices pointing towards climate change, many people are of the belief that climate change is a hoax. Datafluent Inc. is a corporate entity that specializes in mining out actionable insights from raw facts. In this project, we are contracted to use the novel tweets of some individuals to build various Machine Learning (ML) models to identify their beliefs about climate change. These models will be evaluated using the metrics: ***`accuracy, precision, recall, and F1-scores`*** to choose the optimal model. The optimal model’s outcome will help our client to predict how their eco-friendly products will be received by their prospective customers and thus enable them to make strategic business decisions.

## Table of Contents
<a href=#one>1.0 Introduction</a>
- <a href=#1.1>1.1 Project Overview</a>
- <a href=#1.2>1.2 Installing Dependencies and Importing Packages</a>
- <a href=#1.3>1.3 Loading the Data</a>

<a href=#two>2.0 Exploratory Data analysis</a>
- <a href=#2.1>2.1 Data Wrangling</a>

<a href=#three>3.0 Model Building and Evaluations</a>
- <a href=#3.1>3.1 Train-test Split</a>
- <a href=#3.2>3.2 Feature Selection</a>
- <a href=#3.3>3.3 Model Building</a>
- <a href=#3.4>3.4 Model Evaluation and Performance</a>
- <a href=#3.5>3.5 Parameter Tunings</a>

<a href=#four>4.0 Conclusion</a>

<a href=#five>5.0 Submission</a>

<a href=#six>References</a>

<a id="one"></a>

<a id="one"></a>
# 1.0 Introduction
Climate change is a big issue that needs serious attention and many eco-conscious organisations are exploring ways to build eco-friendly products. However, the populace is divided in their beliefs about climate change. **`Datafluent Inc.`** was hired by one such eco-conscious company to build a solution that determines the proportion of the population that believes in climate change. The information that the solution extracts will serve as a basis for the organisation strategic decision-making in building eco-friendly products. In this section, we are going to set the stage for our project on climate change belief analysis of about 16,000 tweets of some individuals. We will also explore the `Machine Learning (ML) tools` we will use in building the sentiment analysis model.

## 1.1 Global Climate Change
Global climate change refers to the average long-term change of climatic conditions over the entire earth. These include warming temperatures and changes in the precipitation as well as the effects of earth's warming such as:

* rising sea levels
* shrinking mountain glaciers
* ice melting at a faster rate than usual in Greenland and Artic
* wildfires
* floods
* droughts
* famine
* increasing desertification etc.

There is a consensus in the scientific community that the earth has been getting warmer in the recent past and this has been linked to human activities, particularly the use of fossil fuels. The dangers of climate change spark debate among people, as we will see when exploring the tweets.  

Earth’s climate has constantly been changing, even long before humans came into the picture. However, scientists have observed unusual changes recently. For example, Earth’s average temperature has been increasing much more quickly than they would expect over the past 150 years as shown in figure 1.1 below. According to scientists, the main human activities that contribute to climate change over the last century are the burning of fossil fuels like coal and oil as well as carbon pollution, which has increased the concentration of atmospheric carbon dioxide(CO2), which is a known greenhouse gas.
<hr>
<div align="center" style="width: 500px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="World_Climate_Change_Chart.png"
     alt="Dummy image 1"
     style="float: center; padding-bottom=0.5em"
     width=900px/>|
    <div align="center">Figure 1.1: World Climate Change Chart</div>
 </div>

## 1.2 Sentiment Analysis
Sentiment analysis is a way of using natural language processing (NLP) and ML tools in identifying opinions expressed in texts. These opinions can be classified as positive (scored as 1), negative (scored as -1), or neutral (scored as 0). Sentiment analysis is important to businesses because it can reveal insights that can be used in improving sales, customer retention, timely problem identification, evaluating customers’ product perceptions, and exploring business opportunities.

In this regard, people can be classified into three based on their perceptions about the causes of climate change. Thus,
1. Pro-sentiment class: they believe that humans contribute to climate change
2. Anti-sentiment class: these are those who do not believe in climate change and those that don’t believe that humans contribute to climate change
3. Neutral class: these are those who take a neutral stand on the matter, displaying no strong positive or negative sentiment.

### *1.2.1 Techniques of sentiment analysis*
Sentiment analysis can be realized in two ways: rule-based and ML sentiment analysis. In rule-based analysis, manual NLP rules like lexicons (lists of words), stemming, tokenization, and parsing are used to simplify complex texts for computers to easily decipher opinions. Whereas, in ML sentiment analysis classification techniques like Naive Bayes, Logistic Regression, Support Vector Machines, Linear Regression, and Deep Learning are used to classify the contextual opinions.

Quality insights are key to data-driven decision-making. At **`Datafluent Inc.`** we strongly believe in extracting only quality insights. Thus, we are going to analyze the given dataset thoroughly using both rule-based and ML techniques of sentiment analysis. This will ensure that no insight is left unturned in informing the decision-making process of our **`esteemed clients.`**

## 1.3 Classification Techniques
We are going to explore the dataset using the following models, compare them and choose the best amongst them.
- Decision Tree Classifier
- Random Forest Classifier
- Linear Support Vector Classifier (LinearSVC)
- Support Vector Classifier (SVC)
- Logistic Regression
- Stochastic Gradient Descent (SGD)

### *1.3.1 Decision tree classifier*
Decision Trees (DTs) are non-parametric supervised learning methods used for classification and regression. Decision trees learn from data to approximate a sine curve with a set of if-then-else decision rules as illustrated in figure 1.2 below. The deeper the tree, the more complex the decision rules, and the fitter the model.

<div align="center" style="width: 500px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="http://www.wordtemplatesdocs.org/wp-content/uploads/2017/11/Decision-Tree-template-free-00004.jpg"
     alt="Decision Tree"
     style="float: center; padding-bottom=0.5em"
     width=900px/>
    <div align="center">Figure 1.2: Dec</div>
 </div>

A decision tree builds classification or regression models in the form of a tree structure. It breaks down data by partitioning it into subsets after each decision while at the same time an associated decision tree is incrementally developed. The final result is a tree with decision nodes and leaf nodes. A decision node has two or more branches. A leaf node represents a classification or decision. The topmost decision node in a tree that corresponds to the best predictor is called the root node. Decision trees can handle both categorical and numerical data.
Decision trees are prone to overfitting. Overfitting happens when the learning algorithm continues to develop hypotheses that reduce training set error at the cost of an increased test set error; One method to tackle overfitting in decision trees is by pruning. There are several approaches used to avoid overfitting in building decision trees namely,
Pre-pruning stops growing the tree earlier, before it perfectly classifies the training set.
Post-pruning allows the tree to perfectly classify the training set, and then post-prune the tree. Practically, the second approach of post-pruning overfits trees is more successful because it is not easy to precisely estimate when to stop growing the tree.
Decision Trees are building blocks for the next machine learning method we will look into, which is the Random Forest Classifier


# Context of the Problem

    
Sentiment analysis refers to the use of natural language processing, text analysis, computational linguistics, and biometrics to systematically identify, extract, quantify, and study affective states and subjective information.
 |
![twitter-bird-animated-logo.gif?w=300&zoom=2](https://dropnerblog.files.wordpress.com/2019/12/twitter-bird-animated-logo.gif?w=300&zoom=2)
    
## Context of the Problem
Companies are constructed around lessening one's environmental impact or carbon footprint. They offer products and services that are environmentally friendly and sustainable, in line with their values and ideals. They would like to determine how people perceive climate change and whether or not they believe it is a real threat. Before a company begins to develop products for a segment of the population that believes in climate change and the effects of climate change on the environment and world, it is logical for such a company to first establish the existence of such demography. Even when it is clear that a section of the population indeed believes that there is a threat of environmental change, we must assess such factors as how much of a threat they believe climate change is. How far they may be willing to go to protect the environment and how passionate they are about supporting the efforts of others to fight climate change.

In this experiment, we aim to explore machine learning as a method to assist us in identifying whether or not a person believes in climate change and ascertain if such a person could be converted to a new customer based on their tweets. We will develop ML models that are capable of classifying tweets leaning positively towards a belief that Climate change is a problem and hence desirous of proferring a solution to this problem or leaning negatively towards that belief, hence treating climatic change as no threat. To produce these models, we must first select an appropriate set of data for training that has been annotated. Then, we pre-process and clean these data, to enable us to generate a feature and target set from them. We will then select relevant ML models, on which to train these datasets and evaluate their performance. Model evaluation is done by comparing model predictions against data that have been annotated but not used for training. Once a satisfactory performance of the model has been achieved, we interpret the patterns learned and apply them for further decision-making in the context of the experiment.

<div align="center" style="width: 950px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://www.itu.int/en/mediacentre/backgrounders/PublishingImages/climate-change-backgrounder.jpg"
     alt="Dummy image 1"
     style="float: center; padding-bottom=0.5em"
     width=950px/>
    </div>

    ### Problem Statement
    Create a Natural Language Processing model to classify whether or not a person believes in climate change, based on their novel tweet data.
    
    ## Comet ML as a tool for Data Science
   <a id="comet"></a>
# Comet ML as a tool for Data Science
<img src="https://d15shllkswkct0.cloudfront.net/wp-content/blogs.dir/1/files/2021/11/cometml.png" width="1000" alt="Drawing" style="width: 1000px;"/>
We will be using Comet to facilitate the model building process as well as a form of version control throughout the development of our model. To begin, we will first install the dependency that allows comet ML to integrate with our notebook and log all changes in our ML model as independent experiments in our comet account. This will allow us to have different versions of our model and will be really helpful in assessing the different properties and functionality of our model versions. This process is done in 3 distinct phases, Pip installation, importation of the experiment, and linking our notebook with our comet account. These processes are completed in the next three cells.

<a id="1.12"></a>
## 1.2 Installing Dependencies and Importing Packages
In order to successfully build the models, there is a need to pip install some dependencies. Thus, we install the *autotime library, Comet, imblearn, and nltk.*

In [3]:
# Installations
%%capture
!pip install ipython-autotime
!pip install comet-ml
!pip install imblearn --user
!pip install --user -U nltk
!pip install wordcloud
%load_ext autotime

time: 33.8 s (started: 2022-06-21 11:39:12 +01:00)


In [ ]:
conda install -c anaconda -c conda-forge -c comet_ml comet_ml

In [10]:
# Libraries for data loading, data manipulation and data visulisation
import numpy as np # for linear algebra
import pandas as pd # for importing, creating and manipulating dataframes

# Visualization Packages
import matplotlib.pyplot as plt
import seaborn as sns
# Warnings
import warnings 
warnings.simplefilter(action="ignore", category=FutureWarning)

# Import comet_ml 
from comet_ml import Experiment

# Packages for text manipulation and Natural language processing
import re
from string import punctuation
import nltk
nltk.download(['stopwords','punkt'])
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize


# Train-test split package
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer,CountVectorizer
from sklearn.utils import resample


# Libraries for data preparation and model building and evaluations
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import RidgeClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier 
from sklearn.ensemble import StackingClassifier
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.model_selection import cross_val_score
from imblearn.pipeline import pipeline
from imblearn.over_sampling import SMOTE

from sklearn import metrics

time: 0 ns (started: 2022-06-21 11:56:10 +01:00)


[nltk_data] Downloading package stopwords to C:\Users\Buhari
[nltk_data]     Shehu\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to C:\Users\Buhari
[nltk_data]     Shehu\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
# Create an experiment with your api key
experiment = Experiment(
    api_key="KQ1UTh7hBvPLWlz3034oIgusG",
    project_name="global-warming-climate-change-sentiment-analysis-zm3",
    workspace="okonp07@-gmail-com",)

<a id="1.3"></a>
## 1.3 Loading the Data
To load your data, first ensure that the raw data and the notebook file are in the same folder on your local machine. The code below will load both the train and test data sets into your notebook. If the files are not in the same folder, you will have to point to the directory in your machine or cloud location where the file is located. After loading your data, it is good practice to call up the loaded data just to verify that the data actually loaded as it should.

In [31]:
#Load the train and test data sets from their respective CSV files
train = pd.read_csv('train.csv')
test = pd.read_csv('test_with_no_labels.csv')

time: 141 ms (started: 2022-06-21 15:09:07 +01:00)


<a id="two"></a>
# 2.0 Exploratory Data analysis

In this section, we are going to objectively analyse the characterisitcs of our train and test data sets, independently, to have a grasp of the features of the data sets.

 Some useful questions: 
* What is the sample size?
* What key- words are useful to establish sentiments?
* What are the sources of the data (News with verifiable sources, Informal tweets, etc )
* Establish sentiments and their percentages in the data (Pro, Anti, Neutral, etc)
* Check for words most commonly featured in the dataset
* Frwequent Hashtags (Pro and anti)
* Consider other conditionalities that will generate insiteful Visuals for you data and use them. 

Remember that EDA is all about visual presentation. Use visuals to tell the Story of your data.

Let's have a glimpse into the datasets and their shapes

In [21]:
# first five rows of the dataset
train.head()

sentiment                                            message  tweetid
0          1  PolySciMajor EPA chief doesn't think carbon di...   625221
1          1  It's not like we lack evidence of anthropogeni...   126103
2          2  RT @RawStory: Researchers say we have three ye...   698562
3          1  #TodayinMaker# WIRED : 2016 was a pivotal year...   573736
4          1  RT @SoyNovioDeTodas: It's 2016, and a racist, ...   466954

time: 407 ms (started: 2022-06-21 14:08:46 +01:00)


In [23]:
# checking the shape of the dataset
train.shape

(15819, 3)

time: 0 ns (started: 2022-06-21 14:09:38 +01:00)


In [24]:
# checking for missing values and data types
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15819 entries, 0 to 15818
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  15819 non-null  int64 
 1   message    15819 non-null  object
 2   tweetid    15819 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 370.9+ KB
time: 782 ms (started: 2022-06-21 14:14:31 +01:00)


We can see from the above that the train dataset consists of 15819 tweets along with their sentiment scores. The dataset is deviod of any missing values.


<a id="2.1"></a>
### 2.1 Data Wrangling
It is essential to clean and preprocess the datasets so as to build more accurate models.The following function will clean and preprocess any tweet parsed into it. 

In [25]:
def tweet_preprocessing(tweet):
    
    '''
    This functions cleans tweets from line breaks, URLs, numbers, etc.
    '''
    
    tweet = tweet.lower() #to lower case
    tweet = tweet.replace('\n', ' ') # remove line breaks
    tweet = tweet.replace('\@(\w*)', '') # remove mentions
    tweet = re.sub(r"\bhttps://t.co/\w+", '', tweet) # remove URLs
    tweet = re.sub('\w*\d\w*', '', tweet) # remove numbers
    tweet = re.sub(r'\#', '', tweet) # remove hashtags. To remove full hashtag: '\#(\w*)'
    tweet = re.sub('\w*\d\w*', '', tweet) # removes numbers?
    tweet = re.sub(' +', ' ', tweet) # remove 1+ spaces

    return tweet

time: 15 ms (started: 2022-06-21 14:45:05 +01:00)


<a id="three"></a>
# 3.0 Model Building and Evaluations
In this section, we are going to build and evalaute various ML models and select the best one amongst them. Will attempt improving the performance of our models by hyper parameter tuning and we will evaluate the performance of these in our dataset:
- Naive Bayes Classifier
- SVC and LinearSVC
- Logistic Regression
- KNN
- Decision Tree
- Random Forests

<a id="3.1"></a>
## 3.1 Train-test Split
After we create a function for preprocessing we must split the data into labels and features (X and y) in order to enable us the run the models on our data sets.

In [26]:
# Splitting the labels and features
train['processed'] = train['message'].apply(tweet_preprocessing)
X = train['processed'].values
y = train['sentiment'].values

time: 1.05 s (started: 2022-06-21 14:49:24 +01:00)


In [27]:
# preprocess testing data by applying our function
test['processed'] = test['message'].apply(tweet_preprocessing)

time: 641 ms (started: 2022-06-21 14:49:25 +01:00)


<a id="3.2"></a>
## 3.2 Feature Selection

<a id="3.3"></a>
## 3.3 Model Building

### a. Naive Bayes Classifier 

In [28]:
# Splitting the labels and fetures into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10,random_state=42,stratify=y)

time: 218 ms (started: 2022-06-21 14:50:34 +01:00)


In [ ]:
mnb = Pipeline([('Count',CountVectorizer()),('classify',MultinomialNB())])
#fitting the model
mnb.fit(X_train, y_train)

#apply model on test data
y_pred_mnb = mnb.predict(X_test)

In [ ]:
# Classification report
print(classification_report(y_test, y_pred_mnb))

As we can see the '-1' and '0' class are poorly predicted when using unbalanced data. Once we implement resampling their f1-score increases for these model but only slightly. While at the same time the overall accuracy is slightly reduced.

### b. SVC and LinearSVC

SVC Provides a best fit to catergorize our data this fit can be nonlinear, while a linearSVC provides a linear interpolation.

In [ ]:
#SVC
svc = Pipeline([('Count',CountVectorizer()),('classify',SVC(max_iter=300,C=1))])

In [ ]:
#linearSVC
linsvc = Pipeline([('Count',CountVectorizer()),('classify',LinearSVC(max_iter=300,C=1))])

### c. Logistic Regression

Models the discrete probability distribution between classes and classifies based on the inflection point of the curve.

In [ ]:
#Logistic Regression
lr = Pipeline([('Count',CountVectorizer()),('classify',LogisticRegression(max_iter=300))])

### d. KNN
The KNN classifier assumes that all data points that similar data points tend to form clusters, close together. It classifies points that are close into the same class.K is the number of neighbours. So K=3 implies we will make our predictions based off of the 3 closest points to the data point beign assessed.

In [ ]:
# Invoke the KNN classifier
knn = Pipeline([('Count',CountVectorizer()),('classify',KNeighborsClassifier(n_neighbors=3))])

### e. Decision Tree

The decision tree uses a tree-like model of decisions and their possible consequences.Starting from the decision itself (called a "node"), each branch of the decision tree represents a possible decision, outcome, or reaction and it works up until there is only one possible outcome left.
<div align="center" style="width: 500px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://www.aihr.com/wp-content/uploads/decision-trees-in-analytics.png"
     alt="Dummy image 1"
     style="float: center; padding-bottom=0.5em"
     width=500px/>

In [ ]:
#Decision Tree
dt = Pipeline([('Count',CountVectorizer()),('classify',DecisionTreeClassifier())])

### f. Random Forest
Using the decision tree as a base estimator,each estimator is trained on a different bootstrap sample having the same size as the training set. At each node of the forest, features are sampled without replacement to increase randomization. Nodes are split to maximise information gain.   

<div align="center" style="width: 500px; font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://cdn.analyticsvidhya.com/wp-content/uploads/2020/05/rfc_vs_dt11.png"
     alt="Dummy image 1"
     style="float: center; padding-bottom=0.5em"
     width=500px/>

In [ ]:
# Call up the Random Forest Sampler
rf = Pipeline([('Count',CountVectorizer()),('classify',RandomForestClassifier())])

<a id="3.4"></a>
### 3.4 Model Evaluation and Performance

In [32]:
num=3
# SVC
scores = cross_val_score(
        svc, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+' SVC models is ' + str(sum(scores)/len(scores)))

NameError: name 'svc' is not defined

time: 156 ms (started: 2022-06-21 15:27:09 +01:00)


In [ ]:
#linearSVC
scores = cross_val_score(
        linsvc, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+ ' LinearSVC models is ' + str(sum(scores)/len(scores)))

In [ ]:
#Logistic Regression
scores = cross_val_score(
        lr, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+' Logistic Regression models is ' + str(sum(scores)/len(scores)))

In [ ]:
#KNN
scores = cross_val_score(
        knn, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+' KNN models is ' + str(sum(scores)/len(scores)))

In [ ]:
#Decision Tree
scores = cross_val_score(
        dt, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+' Decision Tree models is ' + str(sum(scores)/len(scores)))

In [ ]:
#Random Forest
scores = cross_val_score(
        rf, X, y, cv=num, scoring='f1_weighted')
print('The average weighted F1 score over '+str(num)+' KNN models is ' + str(sum(scores)/len(scores)))

The Logistic Regression Model and the LinearSVC model perform the best. The best performance for every model is found when resampling is not done. This could be because because upsampling the minority classes to the level of the majority class results in too much overfitting.

<a id="3.5"></a>
### 3.5 Parameter Tunings

We take a look and see if we can improve our best 2 models: linearSVC and Logistic Regression

In [ ]:
from sklearn.model_selection import GridSearchCV
Cs = [0.001, 0.01, 0.1, 1, 10]
param_grid = {
    'C'     : Cs
    }
grid_SVM = GridSearchCV(LogisticRegression(), param_grid, scoring='f1_weighted', cv=3)
grid_SVM.fit(CountVectorizer().fit_transform(X), y)
grid_SVM.best_params_

In [ ]:
param_grid = {'C'     : Cs }
grid_SVM = GridSearchCV(LinearSVC(), param_grid, scoring='f1_weighted', cv=3)
grid_SVM.fit(CountVectorizer().fit_transform(X), y)
grid_SVM.best_params_

<a id="four"></a>
# 4.0 Conclusion

### Model performance
Several strategies we attempted to improve model performance, ranging from data processing techniques to clean the tweets, data balancing strategies, cross validation and grid search for the best values for model hyperparameters.



### What else we can try
Language models and the use of neural networks were two other strategies that we may implement to see how it will compare with essemble tree based models. It will also provide an opportunity for us to see how neural network models perforn on Natural Language tasks.

### Business case value

The analysis shows generally that the sentiment from the negative tweets arise from a a class of people who believe that climate change and the incidence of global warming is a hoax, or at best overated. Most of these persons defend their ideology with strong resolve and it will therefore be counter productive to try to market environmentally friendly products to them as an effort towards sustainability or to combat the incidence of Climate change. This class are more likely to become customers if other aspects of the product is promoted to them. They are more likely to purchase a product because of quality, fairness in price, use case etc rather than just because it is sustainably produced or good for the environmenmt.

Conversely, People from the other end of the spectrum who display positive sentiments towards climate change definitely believe that climate change is an issue. They show some willingness to "do something" to play their role in combating this issue. What is yet unclear is if their sentiments would translate to any meaningful influence on their product aquisition habbits. They are definitely a better group to target with promotions that highlights the sustainability and environmenmtal friendliness of the products. To be safe though, this message should be embedded in other qualities of the product so that the environmental friendliness would be the Icing on the cake. It would be great for them if they have a product which is good already but also is sustainably produced. 


It will be beneficial for companies or organizations to band together and form groups where Some organisations are mentioned in the tweets, many which share the same values and ideals when it comes to protecting the environment, who have a substantial membership and following on social media of individuals who share the same values and ideals. The formation of potential partnerships with these organisations could lead to brand exposure with individuals who in their daily lives make conscious decisions with regards to the products and services they purchase.

We recommend that the latter strategy of pursuing partnerships with like minded organisations will yield the best results, in terms of finding a group of potential customers who share the same values and ideals, and would be likely to purchase your products and services.

<a id="five"></a>
# Submission

For our final model, we build a stacking classifier to combine Logistic Regression, LinearSVC and Random Forest

In [ ]:
estimators = [
       ('rf', Pipeline([('Count',CountVectorizer(ngram_range=(1,2))),('classify',RandomForestClassifier())])),
         
        ('lnsvc', Pipeline([('Count',CountVectorizer(ngram_range=(1,2))),('classify',LinearSVC(C=0.1))])),
         
        ('MNB',Pipeline([('Count',CountVectorizer()),('classify',MultinomialNB())])),
    
        ('lr', Pipeline([('Count',CountVectorizer(ngram_range=(1,2))),('classify',LogisticRegression(C=1))]))]

In [ ]:
clf = StackingClassifier(
        estimators=estimators
    )

#fitting the model
clf.fit(X, y)

In [ ]:
# End experiment
experiment.end()

In [ ]:
# Display results on comet page
experiment.display()

In [ ]:
# Creating the unseen set, so that we can post to Kaggle and recieve a score based on the performance
x_unseen = test['processed']

submission = pd.DataFrame(
    {'tweetid': test['tweetid'],
     'sentiment': clf.predict(x_unseen)
    })

# save DataFrame to csv file for submission
submission.to_csv("Submission_final.csv", index=False)

# References

<div align="center" style="color: white, background:#000099">
    <b>Copyright &copy 2022 Datafluent Inc.</b>
</div>